_posts/2022-10-07-arrow-parquet-encoding-part-2.md

```diff
@@ -0,0 +1,344 @@
+}
+```
+```json
+{                     <-- "a" is not provided (is null)
+}
+```
+```json
+{                     <-- "a" is non-null but empty
+  "a": []
+}
+```
+```json
+{
+  "a": [null, 2],     <-- "a" has a null and non-null elements
+}
+```
+
+Documents of this format could be stored in this Arrow schema
+
+```text
+Field(name: "a", nullable: true, datatype: List(
+  Field(name: "element", nullable: true, datatype: Int32),
+)
+```
+
+As before, Arrow chooses to represent this in a hierarchical fashion as a `ListArray`. A `ListArray` contains a list of monotonically increasing integers called *offsets*, a validity mask if the list is nullable, and a child array containing the list elements. Each consecutive pair of elements in the offset array identifies a slice of the child array for that index in the ListArray
+
+For example, a list with offsets `[0, 2, 3, 3]` contains 3 pairs of offsets, `(0,2)`, `(2,3)`, and `(3,3)`, and therefore represents a `ListArray` of length 3 with the following values:
+
+```text
+0: [child[0], child[1]]
+1: []
+2: [child[2]]
+```
+
+For the example above with 4 JSON documents, this would be encoded in Arrow as
+
+
+```text
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+                          ┌──────────────────┐ │
+│    ┌─────┐   ┌─────┐    │ ┌─────┐   ┌─────┐│
+     │  1  │   │  0  │    │ │  1  │   │  1  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  0  │   │  1  │    │ │  0  │   │ ??  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  1  │   │  1  │    │ │  1  │   │  2  ││ │
+│    ├─────┤   ├─────┤    │ └─────┘   └─────┘│
+     │  1  │   │  1  │    │ Validity   Values│ │
+│    └─────┘   ├─────┤    │                  │
+               │  3  │    │ child[0]         │ │
+│    Validity  └─────┘    │ PrimitiveArray   │
+                          │                  │ │
+│              Offsets    └──────────────────┘
+     "a"                                       │
+│    ListArray
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+More technical detail is available in the [ListArray format specification](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout).
+
+
+### Parquet Repetition Levels
+
+The example above with 4 JSON documents can be stored in this Parquet schema
+
+```text
+message schema {
+  optional group a (LIST) {
+    repeated group list {
+      optional int32 element;
+    }
+  }
+}
+```
+
+In order to encode lists, Parquet stores an integer *repetition level* in addition to a definition level. A repetition level identifies where in the hierarchy of repeated fields the current value is to be inserted. A value of `0` means a new list in the top-most repeated list, a value of `1` means a new element within the top-most repeated list, a value of `2` means a new element within the second top-most repeated list, and so on.
+
+*Protip*: for the topmost level list, the number of zeros in the `repetition` levels must match the number of rows.
```

Comment:
```suggestion
*Protip*: the number of zeros in the `repetition` levels must match the number of rows.
```
The repetition levels don't belong to a list, they belong to the leaf column

Reply:
I will also try and refrain from changing the style from a more formal tone with things like `*protip*` lol